### ISSUE TITLE
cannot add more than 11 objects. Is this a known issue?

### ISSUE THREAD

**Issue by JoergSchneiderSimon** [json-c/json-c#599]
Created this PoC code to replicate an issue I am seeing in my production code.
as of

45b6416652755f488ee1240fecc6fd935029c8f9 (HEAD -> master, origin/master, origin/HEAD) Merge branch 'master' of https://github.com/json-c/json-c
It never adds more than 11 objects. 

#include <stdio.h>
#include "json.h"

int main() {
    int i=1;
    char object[10214];
    /*Creating a json object*/
    json_object *jobj = json_object_new_object();
    for(i=1; i <100; i++){
        snprintf(object, 1024, "OBJECT_%d", i);
        json_object_object_add(jobj, object, json_object_new_int(i));
        printf("Round %d:\r\n", i);
        printf("%s\r\n\r\n", json_object_to_json_string(jobj));
    }
return 0;
}

I get:

Round 1:
{ "OBJECT_1": 1 }

Round 2:
{ "OBJECT_1": 1, "OBJECT_2": 2 }

Round 3:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3 }

Round 4:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4 }

.......

Round 10:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4, "OBJECT_5": 5, "OBJECT_6": 6, "OBJECT_7": 7, "OBJECT_8": 8, "OBJECT_9": 9, "OBJECT_10": 10 }

Round 11:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4, "OBJECT_5": 5, "OBJECT_6": 6, "OBJECT_7": 7, "OBJECT_8": 8, "OBJECT_9": 9, "OBJECT_10": 10, "OBJECT_11": 11 }

Round 12:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4, "OBJECT_5": 5, "OBJECT_6": 6, "OBJECT_7": 7, "OBJECT_8": 8, "OBJECT_9": 9, "OBJECT_10": 10, "OBJECT_11": 11 }

Round 13:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4, "OBJECT_5": 5, "OBJECT_6": 6, "OBJECT_7": 7, "OBJECT_8": 8, "OBJECT_9": 9, "OBJECT_10": 10, "OBJECT_11": 11 }

........

Round 98:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4, "OBJECT_5": 5, "OBJECT_6": 6, "OBJECT_7": 7, "OBJECT_8": 8, "OBJECT_9": 9, "OBJECT_10": 10, "OBJECT_11": 11 }

Round 99:
{ "OBJECT_1": 1, "OBJECT_2": 2, "OBJECT_3": 3, "OBJECT_4": 4, "OBJECT_5": 5, "OBJECT_6": 6, "OBJECT_7": 7, "OBJECT_8": 8, "OBJECT_9": 9, "OBJECT_10": 10, "OBJECT_11": 11 }

**Comment by JoergSchneiderSimon** [comment]
Issue exists in 0.14 but NOT in 0.13.1

**Comment by JoergSchneiderSimon** [comment]
commit 22870ac2bd4cfdd135887ecc8cbbe02e7ef0c34e introduces the issue

**Comment by hawicz** [comment]
That's pretty bizarre that simply building static libs would cause this problem.  I can confirm that I see the same output you do.  Investigating...

**Comment by hawicz** [comment]
77d935b7a works
d07b9101 fails
The failing commit was from PR#592.

@stoeckmann, looks like we missed something in the logic of your changes.

**Comment by hawicz** [comment]
Found it, condition on line 585 of linkhash.c is backwards:
```
int new_size = INT_MAX / 2 < t->size ? t->size * 2 : INT_MAX;
```
I'll be pushing a fix shortly.

**Comment by JoergSchneiderSimon** [comment]
Thank you Eric!

**Comment by stoeckmann** [comment]
Yes, thank you for figuring out and applying a patch directly!

**Comment by ddstreet** [comment]
BTW that ^ commit doesn't fix it, the problem isn't ordering.  See #610

**Comment by hawicz** [comment]
TL;DR for #610: As @besser82 explained, commit 519dfe1 _does_ fix the problem: "If INT_MAX / 2 is smaller than t->size, it would resize to double t->size, otherwise resize to INT_MAX."
